File: java/L100/LeetCode0103.java
package L100;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

import util.TreeNode;

public class LeetCode0103 {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        boolean left = true;
        while (!queue.isEmpty()) {
            int sz = queue.size();
            LinkedList<Integer> leveList = new LinkedList<>();
            if (left) {
                for (int i = 0;i < sz;++i) {
                    TreeNode cur = queue.poll();
                    leveList.add(cur.val);
                    if (cur.left != null) {
                        queue.add(cur.left);
                    }
                    if (cur.right != null) {
                        queue.add(cur.right);
                    }
                }
            } else {
                for (int i = 0;i < sz;++i) {
                    TreeNode cur = queue.poll();
                    leveList.addFirst(cur.val);
                    if (cur.left != null) {
                        queue.add(cur.left);
                    }
                    if (cur.right != null) {
                        queue.add(cur.right);
                    }
                }
            }
            res.add(leveList);
            left = !left;
        }
        return res;
    }
}

File: cpp/L100/LeetCode0103.cc
#include "../heads/tree_node.h"
#include <deque>
#include <queue>
#include <vector>
using std::vector;

class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (root == nullptr) {
            return res;
        }
        std::queue<TreeNode*> q;
        q.push(root);
        bool left{true};
        while (!q.empty()) {
            int qz = q.size();
            std::deque<int> level;
            for (int i = 0;i < qz;++i) {
                TreeNode *node{q.front()};
                q.pop();
                if (left) {
                    level.push_back(node->val);
                } else {
                    level.push_front(node->val);
                }
                if (node->left != nullptr) {
                    q.push(node->left);
                }
                if (node->right != nullptr) {
                    q.push(node->right);
                }
            }
            left = !left;
            res.emplace_back(vector<int>(level.begin(), level.end()));
        }
        return res;
    }
};

int main() {
    Solution s;
    TreeNode node(3);
    s.zigzagLevelOrder(&node);
}